app/javascript/forms/provider-forms/cloud-provider-form/index.jsx

```diff
@@ -0,0 +1,170 @@
+
+  const loadProviderSchema = (type, newValues = {}) => {
+    API.options(`/api/providers?type=${type}`).then(({ data: { provider_form_schema } }) => { // eslint-disable-line camelcase
+      setState({
+        type,
+        schema: {
+          fields: [
+            ...initialSchema,
+            {
+              component: componentTypes.SUB_FORM,
+              name: type,
+              ...provider_form_schema, // eslint-disable-line camelcase
+            },
+          ],
+        },
+        values: { type, ...newValues },
+      });
+    }).then(miqSparkleOff);
+  };
+
+  useEffect(() => {
+    if (providerId) {
+      miqSparkleOn();
+      API.get(`/api/providers/${providerId}?attributes=endpoints,authentications,zone_name`).then(({
+        type,
+        endpoints: _endpoints,
+        authentications: _authentications,
+        ...provider
+      }) => {
+        const endpoints = _endpoints.reduce((obj, { role, ...endpoint }) => ({
+          ...obj,
+          [role]: endpoint,
+        }), {});
+
+        const authentications = _authentications.reduce((obj, { authtype, ...authentication }) => ({
+          ...obj,
+          [authtype]: authentication,
+        }), {});
+
+        loadProviderSchema(type, { ...provider, endpoints, authentications });
+      });
+    }
+  }, [providerId]);
+
+  const typeSelected = ({ active, values: { type: newType } = {} }) => {
+    if (active === 'type' && type !== newType) {
+      miqSparkleOn();
+      loadProviderSchema(newType);
+    }
+  };
+
+  const onCancel = () => {
+    const message = sprintf(providerId ? __('Edit of %s was cancelled by the user') : __('Add of %s was cancelled by the user'), kind);
+    miqRedirectBack(message, 'success', redirect);
+  };
+
+  const onSubmit = ({ type, ..._data }, { getState }) => {
+    miqSparkleOn();
+
+    const message = sprintf(__('%s %s was saved'), kind, values.name);
+
+    // Retrieve fields from the schema, but omit the validator components as the API doesn't like them
+    const fields = Object.keys(getState().modified).filter(field => !field.match(/^authentications\.[^.]+\.valid$/));
+    // Filter out fields that are not available in the form schema
+    const data = { ..._.pick(_data, fields), ddf: true };
+
+    const request = providerId ? API.patch(`/api/providers/${providerId}`, data) : API.post('/api/providers', { type, ...data });
+
+    request.then(() => miqRedirectBack(message, 'success', redirect)).catch(miqSparkleOff);
+  };
+
+  return (
+    <div>
+      { !providerId && (
+        <MiqFormRenderer
+          schema={typeSelectorSchema}
+          onSubmit={() => undefined}
+          renderFormButtons={() => ''}
+          onStateUpdate={typeSelected}
+        />
+      ) }
+      <EditingContext.Provider value={providerId}>
+        <MiqFormRenderer
+          schema={schema}
+          onSubmit={onSubmit}
+          onCancel={onCancel}
+          initialValues={values}
```

Comment:
Wait second. I know what it does not work. the initialValues is mutable, so you can't change it in runtime. Try unmnout the second form until you have the data based on the provider type.

Reply:
In order to load the initial values, you would have to call `initialize` method form the `formOptions`. We will need a component for that.

Right now, the fields are not filed with values because the were initialized with empty values.

Reply:
This might be a problem, because you should be able to change the provider type (1st form) during runtime.

Reply:
Is it possible to force-remount the component if `values` change?

Reply:
It is not possible nor the right approach. This will need a custom component that will change the values when the type is changed.

Reply:
You can try `initializeOnMount` but it dont think it will help

Reply:
@skateman ok this seems to be possible in a later version of the `final-form` API. We are currently in progress of upgrading to the latest version of final form, but bring some breaking changes. We can test it when we have a pre-release out. That should not take long.

Reply:
I can live with that.

Reply:
I will do some more testing to confirm that the upgrades allow form/field level mutation of initial values.